Code/Framework/AzCore/AzCore/DOM/DomUtils.cpp

```diff
@@ -290,28 +290,33 @@ namespace AZ::Dom::Utils
 
     void* TryMarshalValueToPointer(const AZ::Dom::Value& value, const AZ::TypeId& expectedType)
     {
-        if (!value.IsObject())
+        if (value.IsObject())
         {
-            return nullptr;
-        }
-        auto typeIdIt = value.FindMember(TypeFieldName);
-        if (typeIdIt != value.MemberEnd() && typeIdIt->second.GetString() == PointerTypeName.GetStringView())
-        {
-            if (!expectedType.IsNull())
+            auto typeIdIt = value.FindMember(TypeFieldName);
+            if (typeIdIt != value.MemberEnd() && typeIdIt->second.GetString() == PointerTypeName.GetStringView())
             {
-                auto typeFieldIt = value.FindMember(PointerTypeFieldName);
-                if (typeFieldIt == value.MemberEnd())
-                {
-                    return nullptr;
-                }
-                AZ::TypeId actualTypeId = DomValueToTypeId(typeFieldIt->second);
-                if (actualTypeId != expectedType)
+                if (!expectedType.IsNull())
                 {
-                    return nullptr;
+                    auto typeFieldIt = value.FindMember(PointerTypeFieldName);
+                    if (typeFieldIt == value.MemberEnd())
+                    {
+                        return nullptr;
+                    }
+                    AZ::TypeId actualTypeId = DomValueToTypeId(typeFieldIt->second);
+                    if (actualTypeId != expectedType)
+                    {
+                        return nullptr;
+                    }
                 }
+                return reinterpret_cast<void*>(value[PointerValueFieldName].GetUint64());
             }
-            return reinterpret_cast<void*>(value[PointerValueFieldName].GetUint64());
         }
+        else if (value.IsOpaqueValue())
+        {
+            // the const_cast here isn't pretty, but it's necessary to get to a void* needed in so many places.
+            return AZStd::any_cast<void>(const_cast<AZStd::any*>(&value.GetOpaqueValue()));
```

Comment:
The const_cast isn't necessary here.
This function should accept a AZ::Dom::Value by non-const lvalue reference.

Otherwise this function can be used to return a dangling pointer to a temporary `AZ::Dom::Value`.

For example:

```c++
TryMarshalValueToPointer(AZ::Dom::Value(AZ::Entity{}), azrtti_typeid<AZ::Entity>());
```

Also the opaque pointer case isn't checking against expectedType value at all, where it should be.

Reply:
A better way perhaps to avoid the dangling pointer issue, is to keep this function accepting a `const AZ::Dom::Value&` reference, but instead delete an overload of TryMarshalValueToPointer which accepts a non-const rvalue reference.
```c++
void* TryMarshalValueToPointer(AZ::Dom::Value&& , const AZ::TypeId&) = delete;
```

Reply:
as we discussed, I took your advice and deleted the offending rvalue ref version of the function and changed the opaque path to properly test against the any's type.